torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -956,11 +959,40 @@ void ProcessGroupNCCL::logWorkEnd(WorkNCCL& work) {
       store_, traceKeyEnd_, work.seq_, opTypeToString(work.opType_));
 }
 
+void killNCCLPGWatchDogDeadlock(union sigval /* unused */) {
+  LOG(ERROR) << "NCCL PG Watchdog hang and timeout";
+  int rv = ::kill(::getpid(), SIGKILL);
+  TCCL_THROW_SYSTEM_IF(rv < 0, "kill", errno);
+}
+
 void ProcessGroupNCCL::workCleanupLoop() {
   bool done = false;
 
   std::list<ProcessGroupNCCL::WorkNCCL> completedWorkList;
   while (!done || !terminateProcessGroup_.load()) {
+    // Inside watchdog thread, we might still call CUDA Event and it will
+    // hang the thread and stuck there silent for hours. We want to use a
+    // timer to kill the process if watchdog timeout.
+    struct sigevent action;
+    std::memset(&action, 0, sizeof(action));
+    action.sigev_notify = SIGEV_THREAD;
+    action.sigev_notify_function = killNCCLPGWatchDogDeadlock;
+    timer_t timerId;
+
+    struct itimerspec timerspec;
+    std::memset(&timerspec, 0, sizeof(timerspec));
+    timerspec.it_value.tv_sec = 10 * 60; // Ten minutes and we will need to make this configurable.
+    rv = ::timer_settime(timerId, /*flags=*/0, &timerspec, /*old_value=*/nullptr);
```

Comment:
do we have to keep creating and deleting timer objects?  i wonder if instead we should create a timer_t timer_; once in the class level, and then somewhere inside our workcleanup loop we can keep 'petting' the watchdog by calling `timer_settime` with a fresh time limit?

Reply:
I am not sure if that will work or not.. Let me see, yeah reset it every time will be better. And we can delete the timer in the timeout handler.

Reply:
I think we can refactor this to put the code into functions outside of workCleanupLoop?

init_watchdog_timer  <-- could call this during the PGNccl construction?
pet_watchdog_timer <-- one line call, we could add inside workCleanupLoop?
(kill_watchdog_timer) <-- not really sure why we need this tbh, don't we want it to run forever until the watchdog exits?